### PR TITLE
Speedup roi_perspective_transform op by caching the information of linear interpolation in forward

### PR DESCRIPTION
**Pull request by wanghaoshuang**
test=develop

### CODE REVIEW

File: paddle/fluid/operators/detection/roi_perspective_transform_op.cu [hunk 1]
```diff
@@ -420,100 +441,42 @@ __device__ T get_feature_gradient(T xs, T ys, int w, int h, const int width,
...
-          }
-        }
-      }
+__global__ void RoiTransformGradKernel(int out_size, const int* out2in_idx_data,
+                                       const T* out2in_w_data,
+                                       const T* out_grad_data,
+                                       T* in_grad_data) {
+  CUDA_1D_KERNEL_LOOP(index, out_size * 4) {
+    int in_idx = out2in_idx_data[index];
+    if (in_idx > 0) {
```
Comment (heavengate) [inline paddle/fluid/operators/detection/roi_perspective_transform_op.cu]: should be >= here?

File: paddle/fluid/operators/detection/roi_perspective_transform_op.cu [hunk 2]
```diff
@@ -328,6 +339,16 @@ class CUDAROIPerspectiveTransformOpKernel : public framework::OpKernel<T> {
...
+    auto* out2in_idx = ctx.Output<framework::Tensor>("Out2InIdx");
+    auto* out2in_w = ctx.Output<framework::Tensor>("Out2InWeights");
+
+    int* out2in_idx_data =
+        out2in_idx->mutable_data<int>({out->numel(), 4}, ctx.GetPlace());
+    T* out2in_w_data =
+        out2in_w->mutable_data<T>({out->numel(), 4}, ctx.GetPlace());
+
+    math::SetConstant<platform::CUDADeviceContext, int> init;
+    init(ctx.cuda_device_context(), out2in_idx, static_cast<int>(-1));
```
Comment (heavengate) [inline paddle/fluid/operators/detection/roi_perspective_transform_op.cu:351]: out2in_w 需要初始化么？

File: python/paddle/fluid/tests/unittests/test_roi_perspective_transform_op.py [hunk 3]
```diff
@@ -299,6 +299,10 @@ def test_check_output(self):
         self.check_output()
 
     def test_check_grad(self):
+        self.outputs['Out2InIdx'] = np.zeros(
+            [np.product(self.outputs['Out'].shape), 4]).astype("int32")
+        self.outputs['Out2InWeights'] = np.zeros(
+            [np.product(self.outputs['Out'].shape), 4]).astype("float32")
```
Comment (heavengate) [inline python/paddle/fluid/tests/unittests/test_roi_perspective_transform_op.py:305]: test_roi_pooling里面test_checkout()也会check Intermediate的output，此处是不是可以修改一下

File: paddle/fluid/operators/detection/roi_perspective_transform_op.cu (hunk 2 above)
Comment (wanghaoshuang) [reply paddle/fluid/operators/detection/roi_perspective_transform_op.cu:351]: 根据[这行](https://github.com/PaddlePaddle/Paddle/pull/17090/files#diff-098b08175b9a42797c064d0a85f415c7R450)的条件：
如果out2in_idx[i] == -1, 则out2in_w[i]就不会被用到

所以只需要将out2in_idx中元素初始化为-1即可

File: python/paddle/fluid/tests/unittests/test_roi_perspective_transform_op.py (hunk 3 above)
Comment (wanghaoshuang) [reply python/paddle/fluid/tests/unittests/test_roi_perspective_transform_op.py:305]: roi_pooling的[infer shape](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/roi_pool_op.cc#L35)里有对 Intermediate output的ENFORCE CHECK，所以单测的test_checkout()也需要有Intermediate output。

当前pr只修改了CUDA kernel, CPU kernel的计算没有用到Intermediate output， 所以单测里也没有check.

File: paddle/fluid/operators/detection/roi_perspective_transform_op.cu (hunk 1 above)
Comment (wanghaoshuang) [reply paddle/fluid/operators/detection/roi_perspective_transform_op.cu]: Thx. Fixed.

**Pull request merged**